lib/elixir/lib/module/types/descr.ex

```diff
@@ -495,4 +527,451 @@ defmodule Module.Types.Descr do
+  # `{map_literal, negated_literals}` where `map_literal` is a map type literal
+  # and `negated_literals` is a list of map type literals that are negated from it.
+  #
+  # A map literal is a pair `{:open | :closed, %{keys => value_types}}`.
+  #
+  # For instance, the type `%{..., a: integer()} and not %{b: atom()}` can be represented
+  # by the DNF containing one pair, where the positive literal is `{:open, %{a => integer()}}`
+  # and the negated literal is `{:closed, %{b => atom()}}`.
+  #
+  # The goal of keeping symbolic negations is to avoid distributing difference on
+  # every member of a union which creates a lot of map literals in the union and
+  # requires emptiness checks to avoid creating empty maps.
+  #
+  # For instance, the difference between `%{...}` and `%{a: atom(), b: integer()}`
+  # is the union of `%{..., a: atom(), b: if_set(not integer())}` and
+  # `%{..., a: if_set(not atom()), b: integer()}`; for maps with more keys,
+  # each key in a negated literal may create a new union when eliminated.
+  #
+  # Set-theoretic operations take two DNFs (lists) and return a DNF (list).
+  # Simplifications can be done to prune the latter.
+
+  # Create a DNF from a specification of a map type.
+  defp map_new(open_or_closed, pairs), do: [{open_or_closed, Map.new(pairs), []}]
+
+  defp map_descr(tag, fields), do: %{map: [{tag, fields, []}]}
+
+  @doc """
+  Gets the type of the value returned by accessing `key` on `map`.
+  Does not guarantee the key exists. To do that, use `map_has_key?`.
+  """
+  def map_get!(%{} = descr, key) do
+    if not gradual?(descr) do
+      map_get_static(descr, key)
+    else
+      {dynamic, static} = Map.pop(descr, :dynamic)
+      dynamic_value_type = map_get_static(dynamic, key)
+      static_value_type = map_get_static(static, key)
+      union(intersection(dynamic(), dynamic_value_type), static_value_type)
+    end
+  end
+
+  @doc """
+  Check that a key is present.
+  """
+  def map_has_key?(descr, key) do
+    subtype?(descr, map([{key, term()}], :open))
+  end
+
+  # Assuming `descr` is a static type. Accessing `key` will, if it succeeds,
+  # return a value of the type returned. To guarantee that the key is always
+  # present, use `map_has_key?`. To guarantee that the key may be present
+  # use `map_may_have_key?`. If key is never present, result will be `none()`.
+  defp map_get_static(descr, key) when is_atom(key) do
+    descr_map = intersection(descr, map())
+
+    if empty?(descr_map) do
+      none()
+    else
+      map_split_on_key(descr_map.map, key)
+      |> Enum.reduce(none(), fn typeof_key, union -> union(typeof_key, union) end)
+      |> remove_not_set()
+    end
+  end
+
+  @doc """
+  Check that a key may be present.
+  """
+  def map_may_have_key?(descr, key) do
+    compatible?(descr, map([{key, term()}], :open))
+  end
+
+  # Union is list concatenation
+  defp map_union(dnf1, dnf2), do: dnf1 ++ dnf2
```

Comment:
I wonder if there is something which can be done to prevent these lists from ballooning when calling with similar types (e.g. `union([a], [b])` where one is the subtype of the other could just be `a` or `b`).

```elixir
t = map([a: integer()], :open)
union(t, t) |> union(t) |> to_quoted_string()
"%{..., :a => integer()} or %{..., :a => integer()} or %{..., :a => integer()}"
```

Having long lists will make all other operations expensive and might also become an issue when generating error messages.
That being said, I understand that the structure for maps is quite complex so might be difficult to have a nice normalization step.

Reply:
Sorry still thinking aloud, not sure if the suggestion would work.

Or perhaps with a slightly different representation, we could have all common keys factored out.

e.g.

```elixir
union(
  map([a: integer(), b: integer()], :open),
  map([a: float(), c: float()], :open)
)
```

instead of the current

```elixir
[
  {:open, %{a: %{bitmap: 4}, b: %{bitmap: 4}}, []},
  {:open, %{c: %{bitmap: 8}, a: %{bitmap: 8}}, []}
]
```

would be


```elixir
{
  %{a: %{bitmap: 12},  # new common map - union of possible values for `:a`
  [{:open, %{b: %{bitmap: 4}}, []},  {:open, %{c: %{bitmap: 8}}, []}]  # existing list
}
```

The invariant being that no key can be present in all maps of the list.

Reply:
@sabiwara I think the biggest question is how common those are. We could try to optimize for those cases but, if they are rare, then we are just adding complexity. :(

The other point is that, what happens in this case:

```
union(
  map([a: integer(), b: integer()], :open),
  map([a: float(), c: float()], :open),
  map([b: float(), c: float()], :open),
)
```

Because there is nothing shared between all three, you'd need a tree structure of the shape you provide, which adds a lot of complexity.

That said, I think we should probably apply unions when all of the keys match, but that's how far I would push it for now (either on the union itself OR during pretty printing).

Reply:
> Because there is nothing shared between all three, you'd need a tree structure of the shape you provide, which adds a lot of complexity.

Yes, in this kind of case it should be separate lists. But I'd expect many real world case to be lists of maps with mostly the same keys (e.g. if you have list of ecto structs with each key being sometimes an int/string, sometimes `nil`).